examples/vision/zero_dce.py

```diff
@@ -0,0 +1,461 @@
+        original_pool = tf.nn.avg_pool2d(
+            original_mean, ksize=4, strides=4, padding="VALID"
+        )
+        enhanced_pool = tf.nn.avg_pool2d(
+            enhanced_mean, ksize=4, strides=4, padding="VALID"
+        )
+
+        d_original_left = tf.nn.conv2d(
+            original_pool, self.left_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_original_right = tf.nn.conv2d(
+            original_pool, self.right_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_original_up = tf.nn.conv2d(
+            original_pool, self.up_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_original_down = tf.nn.conv2d(
+            original_pool, self.down_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+
+        d_enhanced_left = tf.nn.conv2d(
+            enhanced_pool, self.left_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_enhanced_right = tf.nn.conv2d(
+            enhanced_pool, self.right_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_enhanced_up = tf.nn.conv2d(
+            enhanced_pool, self.up_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_enhanced_down = tf.nn.conv2d(
+            enhanced_pool, self.down_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+
+        d_left = tf.square(d_original_left - d_enhanced_left)
+        d_right = tf.square(d_original_right - d_enhanced_right)
+        d_up = tf.square(d_original_up - d_enhanced_up)
+        d_down = tf.square(d_original_down - d_enhanced_down)
+        return d_left + d_right + d_up + d_down
+
+
+"""
+### Deep Curve Estimation Model
+
+We implement the Zero-DCE framework as a Keras subclassed model.
+"""
+
+
+def get_enhanced_image(data, output):
+    r1 = output[:, :, :, :3]
+    r2 = output[:, :, :, 3:6]
+    r3 = output[:, :, :, 6:9]
+    r4 = output[:, :, :, 9:12]
+    r5 = output[:, :, :, 12:15]
+    r6 = output[:, :, :, 15:18]
+    r7 = output[:, :, :, 18:21]
+    r8 = output[:, :, :, 21:24]
+    x = data + r1 * (tf.square(data) - data)
+    x = x + r2 * (tf.square(x) - x)
+    x = x + r3 * (tf.square(x) - x)
+    enhanced_image = x + r4 * (tf.square(x) - x)
+    x = enhanced_image + r5 * (tf.square(enhanced_image) - enhanced_image)
+    x = x + r6 * (tf.square(x) - x)
+    x = x + r7 * (tf.square(x) - x)
+    enhanced_image = x + r8 * (tf.square(x) - x)
+    return enhanced_image
+
+
+class ZeroDCE(keras.Model):
+    def __init__(self, **kwargs):
+        super(ZeroDCE, self).__init__(**kwargs)
+        self.dce_model = build_dce_net()
+
+    def compile(self, learning_rate, **kwargs):
+        super(ZeroDCE, self).compile(**kwargs)
+        self.optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
+        self.spatial_constancy_loss = SpatialConsistancyLoss(reduction="none")
+
+    def call(self, data):
+        dce_net_output = self.dce_model(data)
+        return get_enhanced_image(data, dce_net_output)
```

Comment:
This is part of the forward pass of the model so let's make it a method.

In general:

- Something that intrinsically belongs on the model, like the forward pass or something that uses `self`, should be a method
- Something that is standalone and reusable elsewhere should be a utility (like a loss function or a metric)

Reply:
Made `get_enhanced_image` a method instead of a standalone function.